### ISSUE TITLE
Improve scripting support regarding actual output filename

### ISSUE THREAD

**Issue by harrypotterIUP** [yt-dlp/yt-dlp#5471]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a feature unrelated to a specific site
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2022.10.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

If yt-dlp is being called from within another (automatic) script, usually, the output filename is initially unknown. To solve this issue, as far as I am aware, there are currently two options:

- Either, the output filename is set to a specific value and the script continues from there.
- Or there is a way to relay the output filename, that yt-dlp produced, back to the script.

Unfortunately, both options have disadvantages. `-o something` does not guarantee the actual filename to be `something` because post-processing (i.e. merging etc.) might change that to `something.webm` or similar. The same problem affects `--print filename`. An alternative would be to use `--exec`, which has the downside that an additional process has to be spawned just to get the filename. Also, the output has to be parsed to get the filename or reduced to the output filename with `-q`, which means no nice progress bar will show up. Furthermore, platform-independence is lost, because I am not aware of a program that is available on all supported platforms to relay the filename. `--exec echo` is not a good solution because it might interpret escape characters contained in the filename.

Feature requests/suggestions:
- Add note to README.md, because `-o` doesn't mention this so far and I was caught off guard that post-processing changes the filename:
  ```diff
  diff --git a/README.md b/README.md
  --- a/README.md
  +++ b/README.md
  @@ -1168,6 +1166,7 @@ The default location of the .netrc file is `~` (see below).
 
   # OUTPUT TEMPLATE
 
   The `-o` option is used to indicate a template for the output file names while `-P` option is used to specify the path each type of file should be saved to.
  +Note: Due to post-processing (i.e. merging etc.), the actual output filename might differ.
 
   <!-- MANPAGE: BEGIN EXCLUDED SECTION -->
   **tl;dr:** [navigate me to examples](#output-template-examples).
  ```

- Possibility to reliably print post-processing filename on any platform with `--print` without having to use an external tool with `--exec`
- Option `--force-filename` which renames the file after post-processing to whatever `-o` said in the first place (because if the user _wants_ to have an output filename without extension or a wrong extension, why not allow it?)

### Provide verbose output that clearly demonstrates the problem

- [ ] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

**Comment by pukkandan** [comment]
> Also, the output has to be parsed to get the filename or reduced to the output filename with -q, which means no nice progress bar will show up. 

`-qv` will write all log to stderr

> --exec echo is not a good solution because it might interpret escape characters contained in the filename.

The filename is automatically quoted when passed into exec

So this works:
```
❯ yt-dlp test:youtube --remux mkv --exec echo -q
"D:\Programs\Source\yt-dlp\yt-dlp\youtube-dl test video ＂'⧸⧹ä↭𝕐 [BaW_jenozKc].mkv"
```

> Possibility to reliably print post-processing filename on any platform with --print without having to use an external tool with `--exec`

Already exists

```
❯ yt-dlp test:youtube --remux mkv --print after_move:filepath
D:\Programs\Source\yt-dlp\yt-dlp\youtube-dl test video ＂'⧸⧹ä↭𝕐 [BaW_jenozKc].mkv
```

> Option --force-filename which renames the file after post-processing to whatever -o said in the first place (because if the user wants to have an output filename without extension or a wrong extension, why not allow it?)

Related: https://github.com/yt-dlp/yt-dlp/issues/1844

**Comment by harrypotterIUP** [comment]
> `-qv` will write all log/progress to stderr

Sorry, doesn't show a download progress bar for me. But fair enough about stderr.

> The filename is automatically quoted when passed into exec

No, I'm sorry, there is a misunderstanding. The issue is not about quoting the argument passing it to echo. The issue is _within_ echo, because the behavior of echo is "implementation-defined" if the operand contains a backslash. So if the filename contains a backslash, echo might print something different. The only way to reliably print an operand is `printf %s\\n` - which in turn doesn't work on Windows.

> > Possibility to reliably print post-processing filename on any platform with --print without having to use an external tool with `--exec`
> 
> Already exists

Nice. Thank you! I apologize for not realizing `--print after_move:filepath` was possible. :+1: 

> Related: #1844

Yup, true. However, I think two little improvements to README.md will already go a long way until #1844 is resolved:
- Note in the "OUTPUT TEMPLATE" section about `-o` maybe not same as actual filename
- And changing recommendation `--exec echo` to `--print after_move:filepath` because that's a real platform-independent solution that also doesn't need to spin up an extra process

What do you think?

**Comment by pukkandan** [comment]
> > The filename is automatically quoted when passed into exec
> 
> No, I'm sorry, there is a misunderstanding. The issue is not about quoting the argument passing it to echo. The issue is _within_ echo, because the behavior of echo is "implementation-defined" if the operand contains a backslash. So if the filename contains a backslash, echo might print something different. The only way to reliably print an operand verbatim is to use `printf %s\\n` - which in turn doesn't work on Windows.

Ah, I see. That is annoying, but filenames don't have `\` - so a non-issue? But I digress. back to topic:

> * Note in the "OUTPUT TEMPLATE" section about `-o` maybe not same as actual filename. (See also my diff suggestion in the first post.)

This is already there. https://github.com/yt-dlp/yt-dlp/blob/59057343d0d78706a96ca50dcdf339bef1cf244a/README.md#L1305

> * And changing recommendation `--exec echo` to `--print after_move:filepath` because that's a real platform-independent solution that also doesn't need to spin up an extra process.

Will do. I believe the doc was written before `--print after_move:` was a thing

**Comment by MartusDortus** [comment]
- using youtube-dl as alias for yt-dlp
- first line does not work
- second line does work

youtube-dl -x --audio-format mp3 --audio-quality 0 --get-filename -o "${DIR}/%(title)s.%(ext)s" ${@} --exec "chown martus:audio {}";

youtube-dl -q -x --audio-format mp3 --audio-quality 0 -o "${DIR}/%(title)s.%(ext)s" ${@} --exec "chown martus:audio {}" --exec "echo {};";

**Comment by harrypotterIUP** [comment]
@MartusDortus Sorry, not sure what you're trying to say? Are you sure you've posted your comment in the right issue thread?

@pukkandan :+1: +1 to everything you said. Sounds good, thanks for your hard work!

**Comment by pukkandan** [comment]
The documentation changes have been made

> Option --force-filename which renames the file after post-processing to whatever -o said in the first place (because if the user wants to have an output filename without extension or a wrong extension, why not allow it?)

This should be further discussed in #1844

**Comment by harrypotterIUP** [comment]
:heart: Great job, thank you!
